docs/systems.md

```diff
@@ -0,0 +1,126 @@
+# Systems Design Guide
+
+This document describes the design and implementation of the Systems framework in Goose, which enables AI agents to interact with different systems through a unified tool-based interface.
+
+## Core Concepts
+
+### System
+A System represents any component that can be operated by an AI agent. Systems expose their capabilities through Tools and maintain their own state. The core interface is defined by the `System` trait:
+
+```rust
+#[async_trait]
+pub trait System: Send + Sync {
+    fn name(&self) -> &str;
+    fn description(&self) -> &str;
+    fn instructions(&self) -> &str;
+    fn tools(&self) -> &[Tool];
+    async fn status(&self) -> AnyhowResult<HashMap<String, Value>>;
+    async fn call_tool(&self, tool_name: &str, parameters: HashMap<String, Value>) -> ToolResult<Value>;
+}
+```
+
+### Tools
+Tools are the primary way Systems expose functionality to agents. Each tool has:
+- A name
+- A description
+- A set of parameters
+- An implementation that executes the tool's functionality
+
+Tools are defined using a custom attribute macro:
+
+```rust
+#[tool(
+    name = "echo",
+    description = "Echoes back the input"
+)]
+async fn echo(&self, message: String) -> ToolResult<Value> {
+    Ok(json!({ "response": message }))
+}
+```
+
+## Architecture
+
+### Component Overview
+
+1. **System Trait**: The core interface that all systems must implement
+2. **Error Handling**: Specialized error types for tool execution
+3. **Proc Macros**: Simplify tool definition and registration
+
+### Error Handling
+
+The system uses two main error types:
+- `ToolError`: Specific errors related to tool execution
+- `anyhow::Error`: General purpose errors for system status and other operations
+
+This split allows precise error handling for tool execution while maintaining flexibility for general system operations.
+
+## Best Practices
+
+### Tool Design
+
+1. **Clear Names**: Use clear, action-oriented names for tools (e.g., "create_user" not "user")
+2. **Descriptive Parameters**: Each parameter should have a clear description
+3. **Error Handling**: Return specific errors when possible
+4. **Idempotency**: Tools should be idempotent when possible
+5. **State Management**: Be explicit about state modifications
+
+### System Implementation
+
+1. **State Encapsulation**: Keep system state private and controlled
+2. **Error Propagation**: Use `?` operator with `ToolError` for tool execution
+3. **Status Clarity**: Provide clear, structured status information
+4. **Documentation**: Document all tools and their effects
+
+### Example Implementation
+
+Here's a complete example of a simple system:
+
+```rust
+use goose_macros::tool;
+
+struct FileSystem {
+    registry: ToolRegistry,
```

Comment:
what's the ToolRegistry here?

Reply:
in this examply it'd be a private (maybe even entirely hidden) member of the system as a way to implement the macro driven approach. E.g. it'd be a generalization of this from the test system

```
    async fn call_tool(&self, tool_use: ToolUse) -> ToolResult<Value> {
        match tool_use.name.as_str() {
            "echo" => self.echo(tool_use.parameters).await,
            _ => ToolResult::Err(ToolError::ToolNotFound(tool_use.name)),
        }
    }
```